Diff control origin

jcscottiii opened the pull request:
Builds on #49. Closes #51.

After this patch, we can get a diff report including the control origination

``` sh
./fedramp-templater diff fixtures/opencontrols fixtures/FedRAMP_ac-2-1_v2.1.docx
2016/09/06 10:58:38 Read File `fixtures/FedRAMP_ac-2-1_v2.1.docx`
Control: AC-2 (1). Responsible Role in SSP: "OpenControl Role Placeholder". Responsible Role in YAML: "Amazon Elastic Compute Cloud: AWS Staff".
Control: AC-2 (1). Control Origination in SSP: "". Control Origination in YAML: "shared".
2016/09/06 10:58:38 2 diffs detected
```

Also removed one test
From commit msg:

>    The diff logic has now made this test not print out an empty list.
>     Since we don't have complete control over our fixtures without a large
>     new file, let's punt on this for now

Commit:
initial diff of control origination

Commit:
Add logic for control origin diff

Commit:
Remove logic for redundant test

The diff logic has now made this test not print out an empty list.
Since we don't have complete control over our fixtures without a large
new file, let's punt on this for now

Commit:
Fix import from removing test

jcscottiii commented:
@afeld this is ready!

afeld commented on control/summary_table.go:
```diff
@@ -71,6 +73,62 @@ func (st *SummaryTable) Fill(openControlData opencontrols.Data) (err error) {
+	}
+
+	return reports, nil
+}
```
Smaller pieces, please!

afeld commented:
Just my one [broken record] piece of feedback. Otherwise, all good!

Commit:
Split out diff method for control origin into smaller methods

Commit:
Use src mapping

jcscottiii commented:
@afeld this should be ready now!

Commit:
Move fieldSource to infoSource

It makes sense to have infoSource instead of fieldSource because more
than the field could use it such as the controlOriginMapping

Commit:
Run go fmt

jcscottiii commented:
I introduced a new type `infoSoruce` then it made it easier to DRY the code up and just pass in the particular `infoSource`. example of using `infoSource` as an argument:
https://github.com/opencontrol/fedramp-templater/pull/48/files#diff-4977ae1d8e2a7da98d804e096dc5ca92R111

As a result, I changed the `originMapping` type from a struct to a map. This was needed because it made no sense to have a struct with specific member variables `yamlMapping` and `docMapping` when now we have the `infoSource` as a key that can be used to refer instead of having logic to translate from `infoSource` to which member variable, just use `infoSource` out right.

In addition, I created a `field` struct. This made it easy to include where the information came from and easier to create a diff report instead of hardcoding the SSP came first and YAML field came second.

Also, I removed a test, in reality, it added no value.

jcscottiii commented:
Bump

afeld commented on control/summary_table.go:
```diff
@@ -71,27 +73,104 @@ func (st *SummaryTable) Fill(openControlData opencontrols.Data) (err error) {
+func (st *SummaryTable) diffControlOrigination(control string,
+	openControlData opencontrols.Data) ([]reporter.Reporter, error) {
+	// find the control origination section in the document.
+	docControlOriginationData, err := newControlOrigination(st)
```
Minor: maybe this should be a method on the `SummaryTable`?

jcscottiii replied:
i have another idea, what about still have `newControlOrigination`. but put it in the constructor `NewSummaryTable`. `newControlOrigination` it accepts  an `xml.Node` instead of a SummaryTable too. and then add a new member variable to `SummaryTable` `controlOrigination`/ `controlOriginatonTable` that we store upon creation of the overall `SummaryTable`

This way we dont need to remember to call these constructors on things that actually are apart of the SummaryTable but just haven't been assigned to member variables..

afeld commented on control/summary_table.go:
```diff
@@ -71,27 +73,104 @@ func (st *SummaryTable) Fill(openControlData opencontrols.Data) (err error) {
+	// find the control origins noted in the yaml.
+	yamlControlOriginationData := openControlData.GetControlOrigins(control)
+	// find the control origins currently checked in the section in the YAML.
+	yamlControlOrigins := getCheckedOriginsFromYAML(yamlControlOriginationData)
```
Since we need to do this same thing on the origins being returned from the `SummaryTable` _and_ the `yamlControlOriginationData`, maybe that should be its own type with a `GetChecked()` method?

jcscottiii replied:
Done!

afeld commented on control/summary_table.go:
```diff
@@ -71,27 +73,104 @@ func (st *SummaryTable) Fill(openControlData opencontrols.Data) (err error) {
+	onlyInYAMLOrigins := set.Difference(yamlControlOrigins, docControlOrigins)
+	// create the diff report for the origins only in the yaml.
+	onlyInYAMLOriginReports := createControlOriginsDiffReport(onlyInYAMLOrigins, controlOriginMap, control, yamlSrc)
+	reports = append(reports, onlyInYAMLOriginReports...)
```
Maybe wrap the three chunks of code above into a `generateDiffReport()` function that accepts the two `Set`s of origins as arguments?

jcscottiii replied:
i'll leave this as-is. i could move the Difference line inside `createControlOriginsDiffReport` but you would lose variable names like `onlyInDocOrigins` and `onlyInYAMLOrigins` and it would make it harder to follow

afeld commented on control/summary_table.go:
```diff
@@ -71,27 +73,104 @@ func (st *SummaryTable) Fill(openControlData opencontrols.Data) (err error) {
+			}
+			// Get the doc mapping and put it in the doc.
+			reports = append(reports, NewDiff(control, controlOriginationField, firstField, secondField))
+		}
```
Should we be handling the `else`?

jcscottiii replied:
hmm. it shouldn't stop the loop and it's unlikely but will just print a warning out.

jcscottiii replied:
done

afeld commented on control/summary_table_test.go:
```diff
@@ -81,16 +80,4 @@ var _ = Describe("SummaryTable", func() {
-			Expect(diff).To(Equal([]reporter.Reporter{}))
-			Expect(err).To(BeNil())
-		})
-	})
```
You don't think testing this edge case is useful?

jcscottiii replied:
it's useful but the diff won't be empty anymore AC-2 because it'll have a diff with control origin. I'll put it back but put a skip and explain why.

Commit:
Add comment on field struct

Commit:
Add warning for incompatible type

Commit:
Move logic for YAML controlOrigins to opencontrols pkg

Also, restructured the code so that any glue code between doc and yaml
is in the common package

Commit:
Readd test but skip

Commit:
Move call to newControlOrigination to inside NewSummaryTable

jcscottiii commented:
@afeld ready for another look

mogul commented:
Can we get 👀  on this one so we can close it? If so then I think https://github.com/opencontrol/fedramp-templater/issues/51 will be closed (currently malingering `Awaiting Acceptance`).

afeld commented on common/origin/origin.go:
```diff
@@ -0,0 +1,87 @@
+type SrcMapping map[source.Source]string
+
+// IsDocMappingASubstrOf is wrapper that checks if the input string contains the SSP mapping.
+// This is useful because the input string value may have extra characters so we can't do a == (equal to) check.
```
Whitespace, or more than that? If the former, we should just do a [`strings.Trim()`](https://golang.org/pkg/strings/#Trim) on the input.

afeld commented:
Sorry for the delay!
